Send recycle events from pod to pv.

jsafrane opened the pull request:
This allows users to diagnose what's wrong with recycler. Recycler pods are started automatically with a cryptic name and they are deleted immediately when they finish.

e.g, `kubectl describe pv` could show that NFS cannot be mounted (and how many pods have tried it):

```
  FirstSeen     LastSeen        Count   From                            SubobjectPath   Type            Reason          Message
  ---------     --------        -----   ----                            -------------   --------        ------          -------
  59m           59m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(5421800e-347b-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  53m           53m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(3c9809e5-347c-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  46m           46m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(250dd2a2-347d-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  40m           40m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(0d84ea33-347e-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  33m           33m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(f5fb63bf-347e-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  27m           27m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(de7128fd-347f-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  1h            3m              75      {persistentvolume-controller }                  Normal          RecyclerPod     Recycler pod: Successfully assigned recycler-for-nfs to 127.0.0.1
  1h            3m              76      {persistentvolume-controller }                  Normal          RecyclerPod     Recycler pod: Pod was active on the node longer than specified deadline
  1h            1m              12      {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  20m           1m              4       {persistentvolume-controller }                  Warning         RecyclerPod     (events with common reason combined)
```

These steps were necessary:
- added event watcher to volume.RecycleVolumeByWatchingPodUntilCompletion
- pass all these events through volume plugins to volume controller
- rework volume.RecycleVolumeByWatchingPodUntilCompletion unit tests to a table (too much copy-paste)
- fix all unit tests along the way

``` release-note
Send recycle events from pod to pv.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27714)

<!-- Reviewable:end -->

jsafrane commented:
@kubernetes/sig-storage

matchstick commented:
@jsafrane @kubernetes/sig-storage Unless someone says otherwise this should be targeted to 1.3.x or 1.4. Moving to next-candidate and P0.

jsafrane commented:
Rebased

jsafrane commented:
@k8s-bot test this issue: #27460

saad-ali commented:
@pmorie Do you have the bandwidth to review this?

jsafrane commented:
rebased

jsafrane commented:
@k8s-bot test this issue: #30683

jsafrane commented:
@k8s-bot test this issue: #30683

pmorie commented on pkg/volume/util.go:
```diff
@@ -67,34 +66,61 @@ func internalRecycleVolumeByWatchingPodUntilCompletion(pvName string, pod *api.P
+
+			case watch.Deleted:
+				return fmt.Errorf("recycler pod was deleted")
+
```
dead line

pmorie commented on pkg/volume/util.go:
```diff
@@ -106,15 +132,20 @@ type recyclerClient interface {
 	DeletePod(name, namespace string) error
-	WatchPod(name, namespace string, stopChannel chan struct{}) func() *api.Pod
+	WatchPod(name, namespace string, stopChannel chan struct{}) (<-chan watch.Event, error)
+	Event(eventtype, message string)
```
@jsafrane So, `Event` in this interface creates an event for the volume, right?

pmorie replied:
I think it would help if there was godoc in the interface declaration for `WatchPod` and `Event`

jsafrane replied:
godoc added

pmorie commented on pkg/volume/util.go:
```diff
@@ -129,28 +160,63 @@ func (c *realRecyclerClient) DeletePod(name, namespace string) error {
+	podSelector, _ := fields.ParseSelector("metadata.name=" + name)
+	options := api.ListOptions{
+		FieldSelector: podSelector,
+		Watch:         true,
```
off-topic, it seems redundant that you should have to specify `Watch=true` for the watch API endpoint

pmorie commented on pkg/volume/util_test.go:
```diff
@@ -24,106 +24,174 @@ import (
+	name          string
+	existingPod   *api.Pod
+	createPod     *api.Pod
+	eventSequence []watch.Event
```
The difference between `eventSequence` and `expectedEvents` is not clear to me here

jsafrane replied:
`eventSequence` is list of mock events generated by (simulated) recycler Pod, i.e. the test input . `expectedEvents` is list of events that we expect on output, i.e. were sent to to PV that was being deleted.

I'll add some comments around.

pmorie commented on pkg/volume/util_test.go:
```diff
@@ -24,106 +24,174 @@ import (
+			receivedError = err.Error()
+		}
+		if receivedError != test.expectedError {
+			t.Errorf("Test %q failed, expected error %q, got %q", test.name, test.expectedError, receivedError)
```
don't you want to `continue` after this?

jsafrane replied:
ok

pmorie commented on pkg/volume/util_test.go:
```diff
@@ -24,106 +24,174 @@ import (
+			t.Errorf("Test %q failed, expected error %q, got %q", test.name, test.expectedError, receivedError)
+		}
+		if !client.deletedCalled {
+			t.Errorf("Test %q failed, expected deferred client.Delete to be called on recycler pod", test.name)
```
also, continue here?

jsafrane replied:
ok

pmorie commented on pkg/volume/util_test.go:
```diff
@@ -24,106 +24,174 @@ import (
+			expectedError: "recycler pod was deleted",
+		},
+		{
+			// Another recycler pod is already running
```
how about a case where there's already a pod running, and that pod fails?

jsafrane replied:
test added

jsafrane commented:
rebased and fixed issues reported by @pmorie

pmorie commented on pkg/volume/plugins.go:
```diff
@@ -126,7 +126,10 @@ type RecyclableVolumePlugin interface {
 	// NewRecycler creates a new volume.Recycler which knows how to reclaim
 	// this resource after the volume's release from a PersistentVolumeClaim
-	NewRecycler(pvName string, spec *Spec) (Recycler, error)
+	// The recycler will use provided recorder to write any events that might
```
will use the

pmorie commented on pkg/controller/volume/persistentvolume/controller.go:
```diff
@@ -1366,6 +1368,14 @@ func (ctrl *PersistentVolumeController) scheduleOperation(operationName string,
 	}
 }
 
+// newRecyclerEventRecorder return RecycleEventRecorder that sends all events to
```
returns a

pmorie commented:
Whoops, almost LGTM -- a couple little minor comments about... comments.

jsafrane commented:
fixed articles, @pmorie PTAL again

pmorie commented on pkg/volume/plugins.go:
```diff
@@ -124,9 +124,12 @@ type PersistentVolumePlugin interface {
-	// this resource after the volume's release from a PersistentVolumeClaim
-	NewRecycler(pvName string, spec *Spec) (Recycler, error)
+	// NewRecycler creates a new volume.Recycler which knows how to reclaim this
+	// resource after the volume's release from a PersistentVolumeClaim The
```
missing period after PersistentVolumeClaim

jsafrane replied:
fixed

goltermann commented:
@pmorie @jsafrane this needs LGTM soon if it's to make v1.4

jsafrane commented:
rebased, @pmorie, PTAL

jsafrane committed:
Send recycle events from pod to pv.

This allows users to diagnose what's wrong with recycler. Recycler pods are
started automatically with a cryptic name and they are deleted immediately
when they finish.

kubectl describe pods will show:

  FirstSeen     LastSeen        Count   From                            SubobjectPath   Type            Reason          Message
  ---------     --------        -----   ----                            -------------   --------        ------          -------
  59m           59m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(5421800e-347b-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  53m           53m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(3c9809e5-347c-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  46m           46m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(250dd2a2-347d-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  40m           40m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(0d84ea33-347e-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  33m           33m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(f5fb63bf-347e-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  27m           27m             1       {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Unable to mount volumes for pod "recycler-for-nfs_default(de7128fd-347f-11e6-a79b-3c970e965218)": timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  1h            3m              75      {persistentvolume-controller }                  Normal          RecyclerPod     Recycler pod: Successfully assigned recycler-for-nfs to 127.0.0.1
  1h            3m              76      {persistentvolume-controller }                  Normal          RecyclerPod     Recycler pod: Pod was active on the node longer than specified deadline
  1h            1m              12      {persistentvolume-controller }                  Warning         RecyclerPod     Recycler pod: Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "recycler-for-nfs"/"default". list of unattached/unmounted volumes=[vol]
  20m           1m              4       {persistentvolume-controller }                  Warning         RecyclerPod     (events with common reason combined)


These steps were necessary:

- added event watcher to volume.RecycleVolumeByWatchingPodUntilCompletion

- pass all these events through volume plugins to volume controller

- rework volume.RecycleVolumeByWatchingPodUntilCompletion unit tests to a table
  (too much copy-paste)

- fix all unit tests along the way

jsafrane commented:
rebased

pmorie commented:
Sorry for the latency, LGTM.

pwittrock commented:
@saad-ali @pmorie @jsafrane 

FYI, this didn't make 1.4.0

jessfraz commented:
@saad-ali @pmorie do you all need this in 1.4.1?

jessfraz commented:
can you open the PR to cherry-pick it if so

saad-ali commented:
> @saad-ali @pmorie do you all need this in 1.4.1?
> can you open the PR to cherry-pick it if so

@jsafrane @pmorie I'll leave it up to you guys.
